### PR TITLE
Add pyupgrade onto pre-commit

### PR DESCRIPTION
**Pull request by max-sixty**
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

This auto-fixes problems, so the bar to add it on is low...

### CODE REVIEW

File: xarray/core/dataset.py [hunk 1]
```diff
@@ -7011,16 +7001,19 @@ def polyfit(
...
-            int, Tuple[int, int], Mapping[Any, Tuple[int, int]]
-        ] = None,
-        end_values: Union[int, Tuple[int, int], Mapping[Any, Tuple[int, int]]] = None,
+        stat_length: int
+        | tuple[int, int]
+        | Mapping[Any, tuple[int, int]]
+        | None = None,
+        constant_values: (
+            int | tuple[int, int] | Mapping[Any, tuple[int, int]] | None
+        ) = None,
```
Comment (Illviljan) [inline xarray/core/dataset.py:7012]: This is a fun one. Why don't these two inputs have the same format? I definitely prefer the one parenthesis.

File: doc/whats-new.rst [hunk 2]
```diff
@@ -66,6 +66,8 @@ Internal Changes
 - Removed internal checks for ``pd.Panel`` (:issue:`6145`).
   By `Matthew Roeschke <https://github.com/mroeschke>`_.
 
+- Add ``pyupgrade`` pre-commit hook (:issue:`6152`).
```
Comment (Illviljan) [inline doc/whats-new.rst:69]:
```suggestion
- Add ``pyupgrade`` pre-commit hook (:pull:`6152`).
```

**Review by Illviljan — APPROVED** [review]
Looks good to me!

I haven't tried this locally but does it slow down the local runs noticeably? Might be annoying for those who do local pre-commits in that case. But it fits great for the CI and since the bot will autocorrect now I think it outweighs the cons. 
One can always turn off the local pre-commit and rely on the CI instead if it becomes to annoying.

**Comment by keewis** [comment]
> Might be annoying for those who do local pre-commits in that case.

We might need to merge the part of #6024 that deactivates `mypy` in default `pre-commit` runs and add a new CI job instead...

**Comment by max-sixty** [comment]
> does it slow down the local runs noticeably?

These ones are fast! `mypy` is the laggard...

**Pull request merged**